src/System.Private.CoreLib/src/System/Runtime/InteropServices/ComActivator.cs

```diff
@@ -177,6 +197,72 @@ private static Type FindClassType(Guid clsid, string assemblyName, string typeNa
             throw new COMException(string.Empty, CLASS_E_CLASSNOTAVAILABLE);
         }
 
+        private static AssemblyLoadContext GetALC(string assemblyPath)
+        {
+            AssemblyLoadContext alc;
+            return LazyInitializer.EnsureInitialized(ref s_AssemblyLoadContexts).TryGetValue(assemblyPath, out alc) ?
+                alc:
+                AddNewALC(assemblyPath);
+
+            AssemblyLoadContext AddNewALC(string path)
+            {
+                // The static collection of ALCs is readonly. If a new ALC entry needs to be added,
+                // then we must make a copy, add the new ALC, and atomically update the static member.
+                while (true)
+                {
+                    Dictionary<string, AssemblyLoadContext> prevContexts = s_AssemblyLoadContexts;
+                    var curContexts = new Dictionary<string, AssemblyLoadContext>(prevContexts);
+
+                    // Check the new collection prior to adding the new ALC if
+                    // the needed ALC was added in the interim.
+                    AssemblyLoadContext alcMaybe;
+                    if (curContexts.TryGetValue(path, out alcMaybe))
+                    {
+                        return alcMaybe;
+                    }
+
+                    alcMaybe = new ComServerLoadContext(path);
```

Comment:
This can create redundant contexts that will stick around forever. It can be pretty bad if `ComServerLoadContext` takes a while to run and multiple threads are racing through here for the same path.

Do we need a lock here?

Reply:
>It can be pretty bad if `ComServerLoadContext` takes a while to run and multiple threads are racing through here for the same path.

Yes, that is still a concern I couldn't address. If potentially making redundant ALCs is really that much of a concern, I can just use locks for access.

Reply:
> I can just use locks for access

You can still avoid locks for the read path and just use them for updating the dictionary instead of the CAS loop, if it's desirable to avoid locks for reads.

Reply:
>You can still avoid locks for the read path and just use them for replacing the dictionary instead of the CAS loop,

@stephentoub Yep. That is my plan. Will do that tomorrow.